### DISCUSSION TITLE
Is it possible to pass variables into the `script.py.mako` file at revision generation time?

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/alembic#1423] [Usage Questions] [answered]
I'm working on a system where I have a few `x_data` behaviours, and I'd love to be able to express something like this:

```python
# script.mako.py

def upgrade () -> None:
  if "switch" in context.get_x_argument():
    # do stuff conditional on switch
  else:
    # do other stuff

```

as

```mako
# script.mako.py

def upgrade () -> None:
% if some_object.arguments.switch:
  # do stuff
% else
  # do other stuff
% endif
```

Why? Because I'd like to generate the migrations that can then be run in other contexts without having to remember the particular values of the switches passed in at generation time. Yes, I could just copy & paste, but the template engine is *right there*, you know?

TIA!

**Comment by zzzeek** [comment, marked as answer]
sure, you can do whatever you want in the template.

some examples on doing more in the templates you can see for example at https://github.com/sqlalchemy/alembic/blob/8829bd64a70708ef9144604ea0f0a5179d07ccbe/alembic/templates/multidb/script.py.mako#L38

**Reply by ghost** [reply]
Thanks! I guess my next question is, how do I carry that information from invocation time (`alembic revision -m '...'`) to the generation of the migration itself? IOW, there are a bunch of context objects available in `env.py`, and a bunch of hooks, but which ones are available when the mako template is being processed?

**Reply by ghost** [reply]
And I figured it out. There's a dict on the `Config` object called "attributes" that I can use to pass stuff from `env.py` into the mako template. I did this:

```python
# env.py

config.attributes["switch"] = "switch" in context.get_x_arguments()
```

```mako
<% 
switch = config.attributes.get("switch", False)
%>

% if switch:
# do stuff
% else:
# do other stuff
% endif
```

and this works great. Thanks for the pointer!

**Reply by zzzeek** [reply]
right, I knew there was something, thanks for figuring it out

**Reply by andy-slac** [reply]
Sorry for late followup question, could not find answer anywhere. I'm trying to do the same as original poster - pass parameters to make template. I tried to use `config` in the template as the answer suggests but I'm getting an exception `NameError: name 'config' is not defined`. From what I see is that `config` is only defined by default when `autogenerate` is enabled. I also see the option `revision_environment` - do I need to set it to true to have `config` defined? I guess I also need to pass `sql=True` to `revision()`?

**Reply by zzzeek** [reply]
looking at the source code it seems "config" is always in template_args unconditionally so make sure you are on a recent version of alembic and are using the command.revision command to do this render

**Reply by andy-slac** [reply]
Ah, my fault, sorry for the noise. My mistake was that I tried to use `config` inside `<%! ... %>` block. Works as you say in `<% ... %>`.
